pkg/reconciler/nodeconfig_reconciler.go

```diff
@@ -61,95 +68,208 @@ func (ncr *NodeConfigReconciler) reconcileDebounced(ctx context.Context) error {
 		return fmt.Errorf("error listing revisions: %w", err)
 	}
 
-	var revisionToDeploy *v1alpha1.NetworkConfigRevision
-	for i := range revisions.Items {
-		if revisions.Items[i].Status.IsInvalid {
-			continue
-		}
-		revisionToDeploy = &revisions.Items[i]
-		break
-	}
+	revisionToDeploy := getFirstValidRevision(revisions.Items)
 
 	// there is nothing to deploy - skip
 	if revisionToDeploy == nil {
 		return nil
 	}
 
 	nodes, err := listNodes(ctx, ncr.client)
-
 	if err != nil {
 		return fmt.Errorf("error listing nodes: %w", err)
 	}
 
-	if err := ncr.deployNodeConfigs(ctx, nodes, revisionToDeploy); err != nil {
-		return fmt.Errorf("error deploying node configurations: %w", err)
+	nodeConfigs, err := ncr.listConfigs(ctx)
+	if err != nil {
+		return fmt.Errorf("error listing configs: %w", err)
+	}
+
+	available := len(nodes)
+	ready := 0
+	ongoing := 0
+	shouldExit, err := ncr.processConfigs(ctx, nodeConfigs.Items, revisionToDeploy, &ready, &ongoing)
+	if err != nil {
+		return fmt.Errorf("error processing configs: %w", err)
+	}
+	if shouldExit {
+		return nil
+	}
+
+	nodesToDeploy := getNodesToDeploy(nodes, nodeConfigs.Items, revisionToDeploy)
+
+	if err := ncr.updateRevisionCounters(ctx, revisionToDeploy, available, ready, ongoing); err != nil {
+		return fmt.Errorf("error updating revision %s counters: %w", revisionToDeploy.Name, err)
 	}
 
-	revisionToDeploy.Status.IsInvalid = false
-	if err := ncr.client.Status().Update(ctx, revisionToDeploy); err != nil {
-		return fmt.Errorf("error validating revision %s: %w", revisionToDeploy.Name, err)
+	if revisionToDeploy.Status.Ongoing < ncr.updateLimit && len(nodesToDeploy) > 0 {
+		if err := ncr.deployNodeConfig(ctx, nodesToDeploy[0], revisionToDeploy); err != nil {
+			return fmt.Errorf("error deploying node configurations: %w", err)
+		}
 	}
 
 	// remove all but last known valid revision
-	if err := ncr.revisionCleanup(ctx, revisionToDeploy); err != nil {
+	if err := ncr.revisionCleanup(ctx); err != nil {
 		return fmt.Errorf("error cleaning redundant revisions: %w", err)
 	}
 
 	return nil
 }
 
-func (ncr *NodeConfigReconciler) revisionCleanup(ctx context.Context, validRevision *v1alpha1.NetworkConfigRevision) error {
-	revisions, err := listRevisions(ctx, ncr.client)
-	if err != nil {
-		return fmt.Errorf("error listing revisions: %w", err)
+func getFirstValidRevision(revisions []v1alpha1.NetworkConfigRevision) *v1alpha1.NetworkConfigRevision {
+	var result *v1alpha1.NetworkConfigRevision
+	for i := range revisions {
+		if revisions[i].Status.IsInvalid {
+			continue
+		}
+		result = &revisions[i]
+		break
```

Comment:
I think `return &revision[i]` would be easier to read here.

Reply:
Done